File: apps/ios/Sources/Shared/PairingURL.swift
import Foundation

/// `conduit://host[:port]?token=<bearer>` → (endpoint URL, token).
///
/// Lived in the now-deleted `Sources/Views/SettingsSheet.swift` before
/// PR #119's cutover. Hoisted here because ConduitAddServerSheet +
/// ConduitDiscoveryView reference it from `Sources/ConduitUI/Views/`.
enum PairingURL {
    struct Parsed { let endpoint: String; let token: String }

    static func parse(_ raw: String) -> Parsed? {
        guard let components = URLComponents(string: raw),
              let scheme = components.scheme?.lowercased() else { return nil }
        let token = components.queryItems?.first(where: { $0.name.lowercased() == "token" })?.value ?? ""
        guard !token.isEmpty else { return nil }

        if scheme == "conduit", let host = components.host {
            let port = components.port.map { ":\($0)" } ?? ""
            return Parsed(endpoint: "ws://\(host)\(port)", token: token)
        }

        if (scheme == "ws" || scheme == "wss"),
           let host = components.host {
            let port = components.port.map { ":\($0)" } ?? ""
            return Parsed(endpoint: "\(scheme)://\(host)\(port)", token: token)
        }
        return nil
    }
}

File: apps/android/app/src/main/kotlin/sh/nikhil/conduit/PairingURL.kt
package sh.nikhil.conduit

import android.net.Uri

/** Parse pairing URL from QR.
 * Supports:
 *  - `conduit://host[:port]?token=<bearer>`
 *  - `ws[s]://host[:port]?token=<bearer>`
 */
object PairingURL {
    data class Parsed(val endpoint: String, val token: String)

    fun parse(raw: String): Parsed? {
        val uri = runCatching { Uri.parse(raw) }.getOrNull() ?: return null
        val scheme = uri.scheme?.lowercase() ?: return null
        val host = uri.host ?: return null
        val token = uri.getQueryParameter("token").orEmpty()
        if (token.isBlank()) return null
        val port = if (uri.port > 0) ":${uri.port}" else ""
        return when (scheme) {
            "conduit" -> Parsed(endpoint = "ws://$host$port", token = token)
            "ws", "wss" -> Parsed(endpoint = "$scheme://$host$port", token = token)
            else -> null
        }
    }
}
